docs/proposals/20190704-AutomatedUpgrades.md

```diff
@@ -0,0 +1,166 @@
+Since our current `init` code performs tasks in the early boot stage, _and_ it acts is PID 1 once running the `rootfs`, we must always reboot in order run a new version of Talos.
+Aside from these benefits, its just better in general to have more focused scope.
+
+#### The Scope of `init`
+
+Splitting our current `init` into two doesn't completely remove the need for a reboot, but it does reduce the chances that a reboot is required.
+We can decrease these chances by having the new `init` perform only the following tasks:
+
+- mount the special filesystems (e.g. `/dev`, `/proc`, `/sys`, etc.)
+- mount the `rootfs` (a `squashfs` filesystem) using a loopback device
+- start `machined`
+
+By stripping `init` down to the bare minimum to start the `rootfs`, we reduce the chances that we introduce a change that requires upgrading `init`.
+Additionally, on upgrades, we can return control back to `init` to perform a teardown of the node and reutrn it back to a state that is the same as when the machine was freshly booted.
+This provides a clean slate for upgrades.
+
+> TODO: Exactly how `machined` is executed needs to be figured out.
+> We could `switch_root` or we could not.
+
+#### The Scope of `machined`
+
+Once we enter the `rootfs` we have three high level tasks:
+
+- retreive the machineconfig
+- create, format, and mount partitions per the builtin specifications
+- start system and k8s.io services
+
+By having a versioned `machined` that is decoupled from `initramfs` we can encapsulate a version of Talos with one binary that knows how to run Talos.
+It will know how to handle versioned machineconfigs and migrate them to newer formats.
+If we ever decide to change paritition layout, filesystem types, or anything at the blockdevice level, the versioned `machined` will know how to handle that too.
+The same can be said about the services that `machined` is responsible for.
+
+### Running `rootfs` in Memory
+
+To take nodes one step closer to becoming completely ephemeral, we will run the `rootfs` in memory.
+This is part of the solution to remove the need for `ROOT-A` and `ROOT-B` partitions.
+It also allows us to dedicate the owned block device to data that can be reproduced (e.g. containerd data, kubelet data, etcd data, CNI data, etc.).
+
+### Using `squashfs` for the `rootfs`
+
+For performance reasons, and a smaller footprint, we can create our `rootfs` as a `squashfs` image.
+The way in which we can run the `rootfs` is by mounting it using a loopback device.
+Once mounted, we have will perform a `switch_root` into the mounted `rootfs`.
+
+> Note: A benefit of `squashfs` is that it allows us to retain our read only `rootfs` without having to mount it as such.
+
+### Slimming down `rootfs`
+
+To slim down the size of our rootfs, I propose that we remove Kubernetes tarballs entirely and let the `kubelet` download them.
+This is a shift from our immutability goals, but in this case we can embrace configurability at the Kubernetes layer.
```

Comment:
This doesn't really shift immutability.  squashfs is still intrinsically immutable.  We're just decoupling the kubelet image from the rootfs image.  While that does _allow_ a flexion point, it needn't _mandate_ that flexion point.

Reply:
Agree. I can reword.